core/src/main/java/com/yahoo/oak/HashChunk.java

```diff
@@ -0,0 +1,413 @@
+/*
+ * Copyright 2020, Verizon Media.
+ * Licensed under the terms of the Apache 2.0 license.
+ * Please see LICENSE file in the project root for terms.
+ */
+
+package com.yahoo.oak;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+class HashChunk<K, V> extends BasicChunk<K, V> {
+    // HashChunk takes a number of least significant bits from the full key hash
+    // to provide as an index in the EntryHashSet
+    private UnionCodec hashIndexCodec; // to be given
+
+    /*-------------- Members --------------*/
+    private final EntryHashSet<K, V> entryHashSet;
+
+    /*-------------- Constructors --------------*/
+    /**
+     * This constructor is only used when creating the first ever chunk/chunks (without a creator).
+     * The caller might set the creator before returning the HashChunk to the user.
+     *
+     * @param maxItems  is the size of the entries array (not all the entries are going to be in use)
+     *                  IMPORTANT!: it is better to be a power of two,
+     *                  if not the rest of the entries are going to be waisted
+     * @param hashIndexCodec the codec initiated with the right amount of the least significant bits,
+     *                       to be used to get index from the key hash
+     */
+    HashChunk(int maxItems, AtomicInteger externalSize, MemoryManager vMM, MemoryManager kMM,
+        OakComparator<K> comparator, OakSerializer<K> keySerializer,
+        OakSerializer<V> valueSerializer, UnionCodec hashIndexCodec) {
+
+        super(maxItems, externalSize, comparator);
+        assert Math.pow( 2, hashIndexCodec.getFirstBitSize() ) <= maxItems ;
```

Comment:
As general HashChunk can be bigger than 2^bitSize. For example, it can be defined that the "calculated" keyHash will always be even, and odd entries will be for the collisions only. There is a reason to do  bigger chunk and not to enlarge the FirstLevelHashArray. Let's discuss.

Reply:
I will add explanatory comment

Reply:
done